usr.bin/tip/tip.1
.\"	$OpenBSD: tip.1,v 1.43 2014/03/17 20:05:27 tedu Exp $
.\"	$NetBSD: tip.1,v 1.7 1994/12/08 09:31:05 jtc Exp $
.\"
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)tip.1	8.4 (Berkeley) 4/18/94
.\"
.Dd $Mdocdate: March 17 2014 $
.Dt TIP 1
.Os
.Sh NAME
.Nm tip
.Nd serial terminal emulator
.Sh SYNOPSIS
.Nm
.Op Fl nv
.Op Fl Ar speed
.Op Ar system-name
.Sh DESCRIPTION
.Nm
is used to connect to another system over a serial link.
In the era before modern networks, it was typically used to
connect to a modem in order to dial in to a remote host.
It is now frequently used for tasks such as attaching to the
serial console of another machine for administrative or
debugging purposes.
.Pp
The options are as follows:
.Bl -tag -width 4n
.It Fl n
No escape (disable tilde).
.It Fl v
Set verbose mode.
.El
.Pp
If
.Ar speed
is specified, it will override any baudrate specified in the system
description being used.
.Pp
If neither
.Ar speed
nor
.Ar system-name
are specified,
.Ar system-name
will be set to the value of the
.Ev HOST
environment variable.
.Pp
If
.Ar speed
is specified but
.Ar system-name
is not,
.Ar system-name
will be set to a value of 'tip' with
.Ar speed
appended.
For example,
.Ic tip -1200
will set
.Ar system-name
to 'tip1200'.
.Pp
Line access is logged to
.Pa /var/log/aculog .
This file does not exist by default and has to be created
to enable logging.
.Pp
Typed characters are normally transmitted directly to the remote
machine (which does the echoing as well).
A tilde
.Pq Ql ~
appearing as the first character of a line is an escape signal; the
following are recognized:
.Bl -tag -offset indent -width Fl
.It Ic ~^D No or Ic ~.
Drop the connection and exit.
Only the connection is dropped \(en the login session is not terminated.
.It Ic ~c Op Ar name
Change directory to
.Ar name
(no argument implies change to home directory).
.It Ic ~!
Escape to a shell (exiting the shell will return to
.Nm ) .
.It Ic ~\*(Gt
Copy file from local to remote.
.Nm
prompts for the name of a local file to transmit.
.It Ic ~\*(Lt
Copy file from remote to local.
.Nm
prompts first for the name of the file to be sent, then for a command
to be executed on the remote machine.
.It Ic ~p Ar from Op Ar to
Send a file to a remote
.Ux
host.
This command causes the remote
.Ux
system to run the following command string,
sending it the
.Sq from
file:
.Bd -literal -offset indent
stty -echo; cat \*(Gt 'to'; stty echo
.Ed
.Pp
If the
.Sq to
file isn't specified, the
.Sq from
file name is used.
This command is actually a
.Ux
specific version of the
.Ic ~\*(Gt
command.
.It Ic ~t Ar from Op Ar to
Take a file from a remote
.Ux
host.
As in the
.Ic ~p
command, the
.Sq to
file defaults to the
.Sq from
file name if it isn't specified.
The remote host executes the following command string
to send the file to
.Nm :
.Bd -literal -offset indent
cat 'from'; echo '' | tr '\e012' '\e01'
.Ed
.It Ic ~|
Pipe the output from a remote command to a local
.Ux
process.
The command string sent to the local
.Ux
system is processed by the shell.
.It Ic ~$
Pipe the output from a local
.Ux
process to the remote host.
The command string sent to the local
.Ux
system is processed by the shell.
.It Ic ~C
Fork a child process on the local system to perform special protocols
such as XMODEM.
The child program will be run with the following arrangement of
file descriptors:
.Bd -literal -offset indent
0 \*(Lt-\*(Gt remote tty in
1 \*(Lt-\*(Gt remote tty out
2 \*(Lt-\*(Gt local tty stderr
.Ed
.It Ic ~#
Send a
.Dv BREAK
to the remote system.
For systems which don't support the necessary
.Fn ioctl
call, the break is simulated by a sequence of line speed changes and
DEL characters.
.It Ic ~s
Set a variable (see the discussion below).
.It Ic ~v
List all variables and their values (if set).
.It Ic ~^Z
Stop
.Nm
(only available with job control).
.It Ic ~^Y
Stop only the
.Dq local side
of
.Nm
(only available with job control); the
.Dq remote side
of
.Nm ,
the side that displays output from the remote host, is left running.
.It Ic ~?
Get a summary of the tilde escapes.
.El
.Pp
To find the system description, and thus the operating characteristics
of
.Ar system-name ,
.Nm
searches for a system description with a name identical to
.Ar system-name .
The search order is
.Bl -enum -offset indent
.It
If the environment variable
.Ev REMOTE
does not start with a
.Ql /
it is assumed to be a system description, and is considered first.
.It
If the environment variable
.Ev REMOTE
begins with a
.Ql /
it is assumed to be a path to a
.Xr remote 5
database, and the specified database is searched.
.It
The default
.Xr remote 5
database,
.Pa /etc/remote ,
is searched.
.El
.Pp
See
.Xr remote 5
for full documentation on system descriptions.
.Pp
The
.Ar br
capability is used in system descriptions to specify the baud rate
with which to establish a connection.
If the value specified is not suitable, the baud rate to be used may
be given on the command line, e.g.\&
.Ql tip -300 mds .
.Pp
The
.Ar dv
capability is used to specify the device
with which to establish a connection.
For reasons outlined in
.Xr tty 4 ,
.Xr cua 4
devices should be used on architectures which have them.
For those which do not,
.Xr tty 4
devices can be used.
Users in group
.Dq dialer
are permitted to use
.Xr cua 4
devices by default;
permissions on
.Pa /dev/tty00
or
.Pa /dev/ttya
can be changed,
but they will revert to their defaults
after an upgrade or (re)install.
.Pp
When
.Nm
establishes a connection, it sends out the connection message
specified in the
.Ar cm
capability of the system description being used.
.Pp
When
.Nm
prompts for an argument, for example during setup of a file transfer, the
line typed may be edited with the standard erase and kill characters.
A null line in response to a prompt, or an interrupt, will abort the
dialogue and return the user to the remote machine.
.Pp
.Nm
guards against multiple users connecting to a remote system by opening
modems and terminal lines with exclusive access, and by honoring the
locking protocol used by
.Xr uucico .
.Pp
During file transfers
.Nm
provides a running count of the number of lines transferred.
When using the
.Ic ~\*(Gt
and
.Ic ~\*(Lt
commands, the
.Dq eofread
and
.Dq eofwrite
variables are used to recognize end-of-file when reading, and specify
end-of-file when writing (see below).
File transfers normally depend on hardwareflow or tandem mode for flow control.
If the remote system does not support hardwareflow or tandem mode,
.Dq echocheck
may be set to indicate
.Nm
should synchronize with the remote system on the echo of each
transmitted character.
.Ss VARIABLES
.Nm
maintains a set of variables which control its operation.
Some of these variables are read-only to normal users (root is allowed
to change anything of interest).
Variables may be displayed and set through the
.Sq s
escape.
The syntax for variables is patterned after
.Xr vi 1
and
.Xr Mail 1 .
Supplying
.Dq all
as an argument to the set command displays all variables readable by
the user.
Alternatively, the user may request display of a particular variable
by attaching a
.Ql \&?
to the end.
For example,
.Dq escape?
displays the current escape character.
.Pp
Variables are numeric, string, character, or boolean values.
Boolean variables are set merely by specifying their name; they may be
reset by prepending a
.Ql \&!
to the name.
Other variable types are set by concatenating an
.Ql =
and the value.
The entire assignment must not have any blanks in it.
A single set command may be used to interrogate as well as set a
number of variables.
Variables may be initialized at run time by placing set commands
(without the
.Ql ~s
prefix) in the initialization file
.Pa ~/.tiprc ;
the
.Fl v
option additionally causes
.Nm
to display the sets as they are made.
Certain common variables have abbreviations.
The following is a list of common variables, their abbreviations, and
their default values:
.Bl -tag -width Ar
.It Ar baudrate
(num) The baud rate at which the connection was established;
abbreviated
.Ar ba .
.It Ar beautify
(bool) Discard unprintable characters when a session is being
scripted; abbreviated
.Ar be .
.It Ar echocheck
(bool) Synchronize with the remote host during file transfer by
waiting for the echo of the last character transmitted; default is
.Ar off .
.It Ar eofread
(str) The set of characters which signify an end-of-transmission
during a
.Ic ~\*(Lt
file transfer command; abbreviated
.Ar eofr .
.It Ar eofwrite
(str) The string sent to indicate end-of-transmission during a
.Ic ~\*(Gt
file transfer command; abbreviated
.Ar eofw .
.It Ar eol
(str) The set of characters which indicate an end-of-line.
.Nm
will recognize escape characters only after an end-of-line.
.It Ar escape
(char) The command prefix (escape) character; abbreviated
.Ar es ;
default value is
.Ql ~ .
.It Ar exceptions
(str) The set of characters which should not be discarded due to the
beautification switch; abbreviated
.Ar ex ;
default value is
.Dq \et\en\ef\eb .
.It Ar force
(char) The character used to force literal data transmission;
abbreviated
.Ar fo ;
default value is
.Ql ^P .
.It Ar framesize
(num) The amount of data (in bytes) to buffer between filesystem
writes when receiving files; abbreviated
.Ar fr .
.It Ar hardwareflow
(bool) Whether hardware flow control (CRTSCTS) is enabled for the
connection; abbreviated
.Ar hf ;
default value is
.Ql off .
.It Ar host
(str) The name of the host to which you are connected; abbreviated
.Ar ho .
.It Ar linedisc
(num) The line discipline to use; abbreviated
.Ar ld .
.It Ar prompt
(char) The character which indicates an end-of-line on the remote
host; abbreviated
.Ar pr ;
default value is
.Ql \en .
This value is used to synchronize during data transfers.
The count of lines transferred during a file transfer command is based
on receipt of this character.
.It Ar raise
(bool) Upper case mapping mode; abbreviated
.Ar ra ;
default value is
.Ar off .
When this mode is enabled, all lowercase letters will be mapped to
uppercase by
.Nm
for transmission to the remote machine.
.It Ar raisechar
(char) The input character used to toggle uppercase mapping mode;
abbreviated
.Ar rc ;
default value is
.Ql ^A .
.It Ar record
(str) The name of the file in which a session script is recorded;
abbreviated
.Ar rec ;
default value is
.Dq tip.record .
.It Ar script
(bool) Session scripting mode; abbreviated
.Ar sc ;
default is
.Ar off .
When
.Ar script
is
.Li true ,
.Nm
will record everything transmitted by the remote machine in the script
record file specified in
.Ar record .
If the
.Ar beautify
switch is on, only printable
.Tn ASCII
characters will be included in the script file (those characters
between 040 and 0177).
The variable
.Ar exceptions
is used to indicate characters which are an exception to the normal
beautification rules.
.It Ar tabexpand
(bool) Expand tabs to spaces during file transfers; abbreviated
.Ar tab ;
default value is
.Ar false .
Each tab is expanded to 8 spaces.
.It Ar tandem
(bool) Use XON/XOFF flow control to throttle data from the remote host;
abbreviated
.Ar ta .
The default value is
.Ar true
unless the
.Ar nt
capability has been specified in
.Pa /etc/remote ,
in which case the default value is
.Ar false .
.It Ar verbose
(bool) Verbose mode; abbreviated
.Ar verb ;
default is
.Ar true .
When verbose mode is enabled,
.Nm
prints messages while dialing, shows the current number of lines
transferred during a file transfer operations, and more.
.El
.Sh ENVIRONMENT
.Bl -tag -width Fl
.It Ev HOME
The home directory to use for the
.Ic ~c
command.
.It Ev HOST
The default value for
.Ar system-name
if none is specified via the command line.
.It Ev REMOTE
A system description, or an absolute path to a
.Xr remote 5
system description database.
.It Ev SHELL
The name of the shell to use for the
.Ic ~!\&
command; default value is
.Dq /bin/sh .
.El
.Sh FILES
.Bl -tag -width "/var/spool/lock/LCK..*" -compact
.It Pa ~/.tiprc
initialization file
.It Pa tip.record
record file
.It Pa /etc/remote
global
.Xr remote 5
database
.It Pa /var/log/aculog
line access log
.It Pa /var/spool/lock/LCK..*
lock file
.El
.Sh SEE ALSO
.Xr cu 1 ,
.Xr remote 5
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .
.Sh BUGS
The full set of variables is undocumented and should, probably, be
pared down.


usr.bin/tip/tip.c
/*	$OpenBSD: tip.c,v 1.54 2012/12/04 19:31:17 kettenis Exp $	*/
/*	$NetBSD: tip.c,v 1.13 1997/04/20 00:03:05 mellon Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * tip - UNIX link to other systems
 *  tip [-v] [-speed] system-name
 * or
 *  cu phone-number [-s speed] [-l line]
 */

#include <sys/types.h>
#include <sys/socket.h>

#include <util.h>

#include "tip.h"

static void	intprompt(int);
static void	tipin(void);
static int	escape(void);

int
main(int argc, char *argv[])
{
	char *sys = NULL;
	int i, pair[2];

	vinit();

	/* XXX preserve previous braindamaged behavior */
	vsetnum(DC, 1);

	if (strcmp(__progname, "cu") == 0) {
		cumode = 1;
		cumain(argc, argv);
		goto cucommon;
	}

	if (argc > 4) {
		fprintf(stderr, "usage: tip [-nv] [-speed] [system-name]\n");
		exit(1);
	}
	if (!isatty(0)) {
		fprintf(stderr, "%s: must be interactive\n", __progname);
		exit(1);
	}

	for (; argc > 1; argv++, argc--) {
		if (argv[1][0] != '-')
			sys = argv[1];
		else switch (argv[1][1]) {

		case 'v':
			vflag++;
			break;

		case 'n':
			noesc++;
			break;

		case '0': case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
			vsetnum(BAUDRATE, atoi(&argv[1][1]));
			break;

		default:
			fprintf(stderr, "%s: %s, unknown option\n", __progname,
			    argv[1]);
			break;
		}
	}

	(void)signal(SIGINT, cleanup);
	(void)signal(SIGQUIT, cleanup);
	(void)signal(SIGHUP, cleanup);
	(void)signal(SIGTERM, cleanup);
	(void)signal(SIGCHLD, SIG_DFL);

	FD = hunt(sys);
	setbuf(stdout, NULL);

	loginit();
	setparity("none");			/* set the parity table */

	if (ttysetup(vgetnum(BAUDRATE))) {
		fprintf(stderr, "%s: bad baud rate %d\n", __progname,
		    vgetnum(BAUDRATE));
		(void)uu_unlock(uucplock);
		exit(3);
	}
	con();

cucommon:
	/*
	 * From here down the code is shared with
	 * the "cu" version of tip.
	 */

	i = fcntl(FD, F_GETFL);
	if (i == -1) {
		perror("fcntl");
		cleanup(0);
	}
	i = fcntl(FD, F_SETFL, i & ~O_NONBLOCK);
	if (i == -1) {
		perror("fcntl");
		cleanup(0);
	}

	tcgetattr(0, &defterm);
	gotdefterm = 1;
	term = defterm;
	term.c_lflag &= ~(ICANON|IEXTEN|ECHO);
	term.c_iflag &= ~(INPCK|ICRNL);
	term.c_oflag &= ~OPOST;
	term.c_cc[VMIN] = 1;
	term.c_cc[VTIME] = 0;
	defchars = term;
	term.c_cc[VINTR] = term.c_cc[VQUIT] = term.c_cc[VSUSP] =
	    term.c_cc[VDSUSP] = term.c_cc[VDISCARD] =
	    term.c_cc[VLNEXT] = _POSIX_VDISABLE;
	raw();

	(void)signal(SIGALRM, timeout);

	if (vgetnum(LINEDISC) != TTYDISC) {
		int ld = (int)vgetnum(LINEDISC);
		ioctl(FD, TIOCSETD, &ld);
	}

	if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, pair) != 0) {
		(void)uu_unlock(uucplock);
		err(3, "socketpair");
	}

	/*
	 * Everything's set up now:
	 *	connection established (hardwired or dialup)
	 *	line conditioned (baud rate, mode, etc.)
	 *	internal data structures (variables)
	 * so, fork one process for local side and one for remote.
	 */
	printf(cumode ? "Connected\r\n" : "\07connected\r\n");
	tipin_pid = getpid();
	switch (tipout_pid = fork()) {
	case -1:
		(void)uu_unlock(uucplock);
		err(3, "fork");
	case 0:
		close(pair[1]);
		tipin_fd = pair[0];
		tipout();
	default:
		close(pair[0]);
		tipout_fd = pair[1];
		tipin();
	}
	/*NOTREACHED*/
	exit(0);
}

void
con(void)
{
	if (vgetstr(CONNECT) != NULL)
		parwrite(FD, vgetstr(CONNECT), size(vgetstr(CONNECT)));
	logent(vgetstr(HOST), vgetstr(DEVICE), "call completed");
}

void
cleanup(int signo)
{
	(void)uu_unlock(uucplock);
	if (odisc)
		ioctl(0, TIOCSETD, &odisc);
	unraw();
	if (signo && tipout_pid) {
		kill(tipout_pid, signo);
		wait(NULL);
	}
	exit(0);
}

/*
 * put the controlling keyboard into raw mode
 */
void
raw(void)
{
	tcsetattr(0, TCSADRAIN, &term);
}


/*
 * return keyboard to normal mode
 */
void
unraw(void)
{
	if (gotdefterm)
		tcsetattr(0, TCSADRAIN, &defterm);
}

static	jmp_buf promptbuf;

/*
 * Print string ``s'', then read a string
 *  in from the terminal.  Handles signals & allows use of
 *  normal erase and kill characters.
 */
int
prompt(char *s, char *p, size_t sz)
{
	int c;
	char *b = p;
	sig_t oint, oquit;

	stoprompt = 0;
	oint = signal(SIGINT, intprompt);
	oquit = signal(SIGQUIT, SIG_IGN);
	unraw();
	printf("%s", s);
	if (setjmp(promptbuf) == 0)
		while ((c = getchar()) != EOF && (*p = c) != '\n' && --sz > 0)
			p++;
	*p = '\0';

	raw();
	(void)signal(SIGINT, oint);
	(void)signal(SIGQUIT, oquit);
	return (stoprompt || p == b);
}

/*
 * Interrupt service routine during prompting
 */
/*ARGSUSED*/
static void
intprompt(int signo)
{
	(void)signal(SIGINT, SIG_IGN);
	stoprompt = 1;
	printf("\r\n");
	longjmp(promptbuf, 1);
}

/*
 * ****TIPIN   TIPIN****
 */
static void
tipin(void)
{
	int bol = 1;
	int gch;
	char ch;

	/*
	 * Kinda klugey here...
	 *   check for scripting being turned on from the .tiprc file,
	 *   but be careful about just using setscript(), as we may
	 *   send a SIGEMT before tipout has a chance to set up catching
	 *   it; so wait a second, then setscript()
	 */
	if (vgetnum(SCRIPT)) {
		sleep(1);
		setscript();
	}

	while (1) {
		gch = getchar();
		if (gch == EOF)
			cleanup(0);
		gch &= STRIP_PAR;
		if (gch == vgetnum(ESCAPE) && bol) {
			if (!noesc) {
				if (!(gch = escape()))
					continue;
			}
		} else if (!cumode && gch == vgetnum(RAISECHAR)) {
			vsetnum(RAISE, !vgetnum(RAISE));
			continue;
		} else if (gch == '\r') {
			bol = 1;
			ch = gch;
			parwrite(FD, &ch, 1);
			if (vgetnum(HALFDUPLEX))
				printf("\r\n");
			continue;
		} else if (!cumode && gch == vgetnum(FORCE)) {
			gch = getchar();
			if (gch == EOF)
				cleanup(0);
			gch &= STRIP_PAR;
		}
		bol = any(gch, vgetstr(EOL));
		if (vgetnum(RAISE) && islower(gch))
			gch = toupper(gch);
		ch = gch;
		parwrite(FD, &ch, 1);
		if (vgetnum(HALFDUPLEX))
			printf("%c", ch);
	}
}

extern esctable_t etable[];

/*
 * Escape handler --
 *  called on recognition of ``escapec'' at the beginning of a line
 */
static int
escape(void)
{
	int gch;
	esctable_t *p;
	char c = vgetnum(ESCAPE);

	gch = getchar();
	if (gch == EOF)
		cleanup(0);
	gch &= STRIP_PAR;
	for (p = etable; p->e_char; p++)
		if (p->e_char == gch) {
			printf("%s", ctrl(c));
			(*p->e_func)(gch);
			return (0);
		}
	/* ESCAPE ESCAPE forces ESCAPE */
	if (c != gch)
		parwrite(FD, &c, 1);
	return (gch);
}

int
any(int cc, char *p)
{
	char c = cc;
	while (p && *p)
		if (*p++ == c)
			return (1);
	return (0);
}

size_t
size(char *s)
{
	size_t i = 0;

	while (s && *s++)
		i++;
	return (i);
}

char *
interp(char *s)
{
	static char buf[256];
	char *p = buf, c, *q;

	while ((c = *s++)) {
		for (q = "\nn\rr\tt\ff\033E\bb"; *q; q++)
			if (*q++ == c) {
				*p++ = '\\'; *p++ = *q;
				goto next;
			}
		if (c < 040) {
			*p++ = '^'; *p++ = c + 'A'-1;
		} else if (c == 0177) {
			*p++ = '^'; *p++ = '?';
		} else
			*p++ = c;
	next:
		;
	}
	*p = '\0';
	return (buf);
}

char *
ctrl(char c)
{
	static char s[3];

	if (c < 040 || c == 0177) {
		s[0] = '^';
		s[1] = c == 0177 ? '?' : c+'A'-1;
		s[2] = '\0';
	} else {
		s[0] = c;
		s[1] = '\0';
	}
	return (s);
}

/*
 * Help command
 */
void
help(int c)
{
	esctable_t *p;

	printf("%c\r\n", c);
	for (p = etable; p->e_char; p++) {
		printf("%2s", ctrl(vgetnum(ESCAPE)));
		printf("%-2s     %s\r\n", ctrl(p->e_char), p->e_help);
	}
}

/*
 * Set up the "remote" tty's state
 */
int
ttysetup(int speed)
{
	struct termios	cntrl;

	if (tcgetattr(FD, &cntrl))
		return (-1);
	cfsetspeed(&cntrl, speed);
	cntrl.c_cflag &= ~(CSIZE|PARENB);
	cntrl.c_cflag |= CS8;
	if (vgetnum(DC))
		cntrl.c_cflag |= CLOCAL;
	if (vgetnum(HARDWAREFLOW))
		cntrl.c_cflag |= CRTSCTS;
	cntrl.c_iflag &= ~(ISTRIP|ICRNL);
	cntrl.c_oflag &= ~OPOST;
	cntrl.c_lflag &= ~(ICANON|ISIG|IEXTEN|ECHO);
	cntrl.c_cc[VMIN] = 1;
	cntrl.c_cc[VTIME] = 0;
	if (vgetnum(TAND))
		cntrl.c_iflag |= IXOFF;
	return (tcsetattr(FD, TCSAFLUSH, &cntrl));
}

static char partab[0200];

/*
 * Do a write to the remote machine with the correct parity.
 * We are doing 8 bit wide output, so we just generate a character
 * with the right parity and output it.
 */
void
parwrite(int fd, char *buf, size_t n)
{
	int i;
	char *bp;

	bp = buf;
	if (bits8 == 0)
		for (i = 0; i < n; i++) {
			*bp = partab[(*bp) & 0177];
			bp++;
		}
	if (write(fd, buf, n) < 0) {
		if (errno == EIO)
			tipabort("Lost carrier.");
		/* this is questionable */
		abort();;//
		perror("write");
	}
}

/*
 * Build a parity table with appropriate high-order bit.
 */
void
setparity(char *defparity)
{
	int i, flip, clr, set;
	char *parity;
	extern const unsigned char evenpartab[];

	if (vgetstr(PARITY) == NULL)
		vsetstr(PARITY, defparity);
	parity = vgetstr(PARITY);
	if (strcmp(parity, "none") == 0) {
		bits8 = 1;
		return;
	}
	bits8 = 0;
	flip = 0;
	clr = 0377;
	set = 0;
	if (strcmp(parity, "odd") == 0)
		flip = 0200;			/* reverse bit 7 */
	else if (strcmp(parity, "zero") == 0)
		clr = 0177;			/* turn off bit 7 */
	else if (strcmp(parity, "one") == 0)
		set = 0200;			/* turn on bit 7 */
	else if (strcmp(parity, "even") != 0) {
		(void) fprintf(stderr, "%s: unknown parity value\r\n", parity);
		(void) fflush(stderr);
	}
	for (i = 0; i < 0200; i++)
		partab[i] = ((evenpartab[i] ^ flip) | set) & clr;
}
